src/compiler/compiler.rs

```diff
@@ -921,129 +944,137 @@ where
         //
         // The check for the `CARGO` env acts as a guardrail against false positives.
         // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
-        args.iter().next().and_then(|arg1| {
+        if let Some(arg1) = args.iter().next() {
             if is_rustc_like(arg1) {
-                Some(PathBuf::from(arg1))
-            } else {
-                None
+                must_be_rust = true;
+                rustc_executable = PathBuf::from(arg1)
             }
-        })
-    } else {
-        None
+        }
     };
 
-    let creator1 = creator.clone();
     let pool = pool.clone();
+    if !must_be_rust && is_like_c_compiler(&rustc_executable) {
+        let executable = executable.to_owned();
+        let cc = detect_c_compiler(creator, executable, args, env.to_vec(), pool).await;
+        return cc.map(|c| (c, None));
+    }
+
+    // Even if it does not look like rustc like it might still be rustc driver
+    // so we do full check
+    let creator1 = creator.clone();
     let cwd = cwd.to_owned();
-    if let Some(rustc_executable) = rustc_executable {
-        // Sanity check that it's really rustc.
-        let mut child = creator.clone().new_command_sync(executable);
-        // We're wrapping rustc if the executable doesn't match the detected rustc_executable. In this case the wrapper
-        // expects rustc as the first argument.
-        if rustc_executable != executable {
-            child.arg(&rustc_executable);
-        }
+    let mut child = creator.clone().new_command_sync(executable);
+    // We're wrapping rustc if the executable doesn't match the detected rustc_executable. In this case the wrapper
+    // expects rustc as the first argument.
+    if rustc_executable != executable {
+        child.arg(&rustc_executable);
+    }
 
-        child.env_clear().envs(ref_env(env)).args(&["-vV"]);
+    child.env_clear().envs(ref_env(env)).args(&["-vV"]);
 
-        let rustc_vv = run_input_output(child, None).await.map(|output| {
-            if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
-                if stdout.starts_with("rustc ") {
-                    return Ok(stdout);
-                }
+    let rustc_vv = run_input_output(child, None).await.and_then(|output| {
+        if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
+            if stdout.starts_with("rustc ") {
+                return Ok(stdout);
             }
-            Err(ProcessError(output))
-        })?;
+        }
+        Err(ProcessError(output)).context("Failed to launch subprocess for compiler determination")
+    });
 
-        match rustc_vv {
-            Ok(rustc_verbose_version) => {
-                let rustc_executable2 = rustc_executable.clone();
+    // If it does not look like rustc and it's verification failed check for c compiler
+    match rustc_vv {
```

Comment:
same, could you please try to move this into a function

Reply:
I moved both into one function as both are tightly coupled.